std/stdio.d

```diff
@@ -4097,17 +4143,17 @@ private struct ReadlnAppender
     }
     void putonly(char[] b)
     {
+        import std.algorithm: max;
         assert(pos == 0);   // assume this is the only put call
-        if (b.length > cap)
+        if (b.length > max(buf.length, buf.capacity))
```

Comment:
How about extracting a `reserveWithoutAllocating` from `reserve` that returns false instead of allocating a new buffer. And then change `putonly` to this:

```d
    void putonly(char[] b)
    {
        assert(pos == 0);   // assume this is the only put call
        if (reserveWithoutAllocating(b.length))
            memcpy(buf.ptr + pos, b.ptr, b.length);
        else
            buf = b.dup;
        pos = b.length;
    }
```

No `.capacity` call when `buf.length` is sufficient, at most one `.capacity` call. Tight allocation with `putonly` (not sure if this is important, but it seems to be the point of `putonly`), spacious allocation with `putchar`.